app/dojo.py

```diff
@@ -0,0 +1,279 @@
+        subheading_allocations_offices = "\nROOM NAME: %s\tROLE\t\tACCOMODATION\n"
+        heading_allocations_livingspaces = "\nALLOCATIONS - LIVING SPACES"
+        sub_heading_allocations_livingspaces = "\nROOM NAME: %s\tROLE\t\tOFFICE\n"
+
+        if origin == "print_allocations":
+            return(heading_allocations_offices, subheading_allocations_offices,
+                heading_allocations_livingspaces, sub_heading_allocations_livingspaces)
+
+        #if origin is not print_allocations, it is print_unallocated
+        #headings and sub_headings for print_unallocations.
+        heading_unallocated_offices = "\nUNALLOCATED - OFFICES"
+        heading_unallocated_livingspaces = "\nUNALLOCATED - LIVING SPACES"
+        sub_heading_unallocated = "\nPERSON NAME\t\tROLE\n"
+
+        return(heading_unallocated_offices, heading_unallocated_livingspaces,
+         sub_heading_unallocated)
+
+    def print_allocations(self, filename=""):
+        headings = self.choose_headings("print_allocations")
+        heading_office = headings[0]
+        sub_heading_office = headings[1]
+        heading_livingspace = headings[2]
+        sub_heading_livingspace = headings[3]
+
+        cprint(heading_office,  "blue")
+        for room in self.rooms:
+            if room.category == "Office":
+                cprint(sub_heading_office % room.name.upper(), "yellow")
+                for occupant in room.occupants:
+                    print("%s \t\t%s\t\t%s" %(str(occupant), 
+                        occupant.category.upper(), str(occupant.accomodation)))
+
+        cprint(heading_livingspace, "blue")
+        for room in self.rooms:
+            if room.category == "Living Space":
+                cprint(sub_heading_livingspace % room.name.upper(),  "yellow")
+                for occupant in room.occupants:
+                    print("%s \t\t%s\t\t%s" %(str(occupant), 
+                        occupant.category.upper(), str(occupant.office)))
+                            
+        if filename:
+            with open(filename, "w") as outputfile:
+                outputfile.write(heading_office)
+                for room in self.rooms:
+                    if room.category == "Office":
+                        outputfile.write(sub_heading_office % room.name.upper())
+                        for occupant in room.occupants:
+                            outputfile.write("%s \t\t%s\t\t%s\n" %(str(occupant), 
+                        occupant.category.upper(), str(occupant.accomodation)))
+
+                outputfile.write(heading_livingspace)
+                for room in self.rooms:
+                    if room.category == "Living Space":
+                        outputfile.write(sub_heading_livingspace % room.name.upper())
+                        for occupant in room.occupants:
+                            outputfile.write("%s \t\t%s\t\t%s\n" % (str(occupant), 
+                        occupant.category.upper(), str(occupant.office)))
+
```

Comment:
From L239-244 you could have a `elif room.category == "Living Space"` in L237 instead of looping through the whole list again.